### DISCUSSION TITLE
HeatIt Thermostat - Formatting Setpoint Commands

### DISCUSSION THREAD

**Discussion by Rollem1** [zwave-js/node-red-contrib-zwave-js#163] [Q&A] [answered]
Hi Everyone,

Firstly, I've just switched over to using zwave-js and just wanted to say thanks for this awesome project. It's so much easier to manage my Z Wave nodes than in OZW in node red!

I have managed to get most things working however struggling with a Thermofloor HeatIt thermostat with which I have two issues. 

1. Although I am able to read back the current temp, heatmode (there are three - Heat, Energy Save Heating (the setback temperature) and Off), the Heat Setpoint and the Energy Save Heating setpoint - I am unable to get the value to poll straight after it is set either using command factory or in the UI. The device does send the updated state eventually but it can take a few minutes to update. Is there a way of forcing this? I have read in the wiki about the forceUpdate command but I am not sure how to use it correctly.

![image](https://user-images.githubusercontent.com/45107860/145729044-8bfc9eda-6534-4027-8bc7-31c2a93d6ec2.png)

2. I am unable to set either the Heat or Heat Eco Setpoints using command factory. I can set them no problem from the UI. Looking at the command sent in UI Monitor they look like this:-

For Heating Mode
![image](https://user-images.githubusercontent.com/45107860/145729085-a674dde8-7199-4474-bae3-f0c1673b7ea1.png)

and for Energy Save Heating Mode
![image](https://user-images.githubusercontent.com/45107860/145729105-4148b29f-65d2-4aee-b1f5-e2acc5bc0d78.png)

To try it - I am injecting a payload of 20.5 (or some value) and a topic of 3 (node address) but I think because there are two setpoints I have to add more parameters. However when I do I hit errors in my formatting such as "Error: Key in object structure must evaluate to a string; got: undefined"

I would like to know how to format the command factory command correctly but also how to inject the whole message from a function because ultimately I have buttons on my UI that set the temp up / down by 0.5 degree each time the relevant button is pressed so this is my ultimate goal. 

Jon

**Comment by marcus-j-davies** [comment]
Hi @Rollem1,

First be aware CCAPI and ValueAPI act very differently from each other.
if you want instant feedback, its always best to use the ValueAPI.

The value API will poll the value after it is set.
CMD Factory supports the ValueAPI also - and the ValueAPI is what is used on the UI

The CCAPI is an extremely  manual API, you have to do everything yourself - whilst easier to look at, is a very lazy API.
The UI tab uses the ValueAPI as it updates the value locally.

With that said, setting setpoints (with the different types) via the CCAPI, your params should evaluate to something like.
`["Heating", 22, 0x00]`  // Setpoint Type, Value, Scale (0x00 - Celsius, 0x01 - Fahrenheit)

If you click **View Parameter Definition** - it gives some indication on what is expected.

**Setpoint types.**
"N/A" = 0x00,
"Heating" = 0x01, // CC v1
"Cooling" = 0x02, // CC v1
"Furnace" = 0x07, // CC v1
"Dry Air" = 0x08, // CC v1
"Moist Air" = 0x09, // CC v1
"Auto Changeover" = 0x0a, // CC v1
"Energy Save Heating" = 0x0b, // CC v2
"Energy Save Cooling" = 0x0c, // CC v2
"Away Heating" = 0x0d, // CC v2
"Away Cooling" = 0x0e, // CC v3
"Full Power" = 0x0f, // CC v3


as for **forceUpdate** if using ValueAPI - its of no use.
but if you want to continue with CCAPI, try this for forceUdpate

`{"property":"setpoint", "propertyKey":1 }` // for "Heating"

**Comment by marcus-j-davies** [comment, marked as answer]
Full example CCAPI

```javascript
let Message = {
    payload: {
        mode: "CCAPI",
        node: 2,
        cc: "Thermostat Setpoint",
        method: "set",
        params: ["Heating", 22, 0x00],
        forceUpdate: {
            property: "setpoint",
            propertyKey: 1
        }
    }
}
return Message
```

Full example ValueAPI

```javascript
let Message = {
    payload: {
        mode: "ValueAPI",
        node: 2,
        method: "setValue",
        params: [<ValueID>, <Value>],
   
    }
}
return Message
```

**Reply by marcus-j-davies** [reply]
You can obtain the ValueID (if using the ValueAPI)- by double clicking a values title in the UI.
![image](https://user-images.githubusercontent.com/55892693/145732003-bfe9771b-ab35-4f74-ab4e-4f66520debf6.png)

**Reply by Rollem1** [reply]
Thanks for the quick reply Marcus,

I went with the ValueAPI option and confirm that the value is updating straight after the change.

Jon
